### PR TITLE
HPCC-35958 Improve the unit tests for seeking into compressed files

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
Added to help verify if there is a problem with seeked reading from compressed files.  No problems found.

File: testing/unittests/filetests.cpp [hunk 1]
```diff
@@ -471,4 +471,129 @@ CPPUNIT_TEST_SUITE_REGISTRATION( JlibStreamTest );
...
+    {
+        unsigned verifySize = 0x1000; // 4KB
+        unsigned numReads = 2000;
+        testSeeks(compression, verifySize, numReads);
+    }
+    void testUncompressedSeeks()
+    {
+        testSeeks(COMPRESS_METHOD_NONE);
+        testSeeks(COMPRESS_METHOD_NONE, 0x400, 1000000);
+    }
```
Comment (AttilaVamos) [inline testing/unittests/filetests.cpp:581]: Do we have some estimation about the runtime of this test?

File: testing/unittests/filetests.cpp [hunk 2]
```diff
@@ -471,4 +471,129 @@ CPPUNIT_TEST_SUITE_REGISTRATION( JlibStreamTest );
...
+    {
+        Owned<IFile> file(createIFile(testFilename));
+        Owned<IFileIO> io;
+        if (compression != COMPRESS_METHOD_NONE)
+            io.setown(createCompressedFileReader(file, nullptr, useDefaultIoBufferSize, false, IFEnone));
+        else
+            io.setown(file->open(IFOread));
+
+        // Perform random reads and verify data
+        std::mt19937 randomGenerator; // Use the default seed, do not re-initialize
```
[the remark this answers is not in the document]
Comment (dcamper) [reply testing/unittests/filetests.cpp]: This is a valid comment, but it may be that you want a deterministic "random" path through the file. If you want a random position then use something simple like `mt19937{random_device()()}`.

File: testing/unittests/filetests.cpp [hunk 3]
```diff
@@ -471,4 +471,129 @@ CPPUNIT_TEST_SUITE_REGISTRATION( JlibStreamTest );
...
+
+            // Fill buffer with semi-random data based on offset
+            for (size32_t i = 0; i < thisSize; i++)
+            {
+                offset_t pos = offset + i;
+                // Simple algorithm: multiply by prime, add high byte to create pseudo-randomness
+                buffer[i] = (byte)((pos * 31 + pos / 256) & 0xFF);
+            }
+
+            io->write(offset, thisSize, buffer);
```
Comment (dcamper) [inline testing/unittests/filetests.cpp:506]: Don't you need to capture the bytes actually written?

```
thisSize =  io->write(offset, thisSize, buffer);
```

File: testing/unittests/filetests.cpp [hunk 4]
```diff
@@ -471,4 +471,129 @@ CPPUNIT_TEST_SUITE_REGISTRATION( JlibStreamTest );
...
+
+        // Perform random reads and verify data
+        std::mt19937 randomGenerator; // Use the default seed, do not re-initialize
+        MemoryAttr buffer(verifySize);
+
+        for (unsigned i = 0; i < numReads; i++)
+        {
+            offset_t offset = randomGenerator() % (size - verifySize);
+
+            io->read(offset, verifySize, buffer.mem());
```
[the remark this answers is not in the document]
Comment (dcamper) [reply testing/unittests/filetests.cpp]: An assert would be a good idea, I think.

File: testing/unittests/filetests.cpp [hunk 5]
```diff
@@ -471,4 +471,129 @@ CPPUNIT_TEST_SUITE_REGISTRATION( JlibStreamTest );
...
+        testSeeks(COMPRESS_METHOD_LZ4);
+        testSeeks(COMPRESS_METHOD_LZ4HC3);
+        testSeeks(COMPRESS_METHOD_ZSTD);
+        testSeeks(COMPRESS_METHOD_LZW);
+        testSeeks(COMPRESS_METHOD_LZ4, 0x400, 100000); // Test smaller reads and more of them to increase chance of hitting edge cases in the compressed seek logic
+    }
+    void cleanup()
+    {
+        Owned<IFile> file(createIFile(testFilename));
+//        file->remove();
```
[the remark this answers is not in the document]
Comment (dcamper) [reply testing/unittests/filetests.cpp]: This seems valid. Each test overwrites the file, but I don't think anything cleans it up at the end.

**Review by dcamper — CHANGES_REQUESTED** [review]
A few comments.

File: testing/unittests/filetests.cpp (hunk 1 above)
Comment (ghalliday) [reply testing/unittests/filetests.cpp:581]: A few seconds - I'll need to look.  But I adjusted the timings so it completed in a few minutes.  (Compressed is much slower.)

**Comment by ghalliday** [comment]
It took ~150s on my desktop.

**Review by dcamper — APPROVED** [review]
Looks good.
